File: ts/src/pro/test/Exchange/test.watchOrderBookForSymbols.ts

import testOrderBook from '../../../test/Exchange/base/test.orderBook.js';
import testSharedMethods from '../../../test/Exchange/base/test.sharedMethods.js';
import { InvalidNonce } from '../../../base/errors.js';
import { Exchange, OrderBook } from '../../../../ccxt.js';

async function testWatchOrderBookForSymbols (exchange: Exchange, skippedProperties: object, symbols: string[]) {
    const method = 'watchOrderBookForSymbols';
    // as in `watchOrderBook`, a pending subscription can not be cancelled, so the
    // loop has to be bounded by the deadline alone. waiting for every requested
    // symbol to be seen would hang forever whenever one of them stays idle.
    const maxIdleTime = 5000;
    let currentTime = exchange.milliseconds ();
    const deadline = currentTime + 15000;
    let idle = false;
    while ((currentTime < deadline) && !idle) {
        let response: OrderBook | undefined = undefined;
        let succeeded = true;
        const startTime = exchange.milliseconds ();
        try {
            response = await exchange.watchOrderBookForSymbols (symbols);
        } catch (e) {
            // interim workaround for InvalidNonce raised by the c# runtime
            if (!testSharedMethods.isTemporaryFailure (e) && !(e instanceof InvalidNonce)) {
                throw e;
            }
            succeeded = false;
        }
        currentTime = exchange.milliseconds ();
        if ((succeeded === true) && (response !== undefined)) {
            testOrderBook (exchange, skippedProperties, method, response, undefined);
            testSharedMethods.assertInArray (exchange, skippedProperties, method, response, 'symbol', symbols);
            const elapsed = currentTime - startTime;
            if (elapsed > maxIdleTime) {
                idle = true;
            }
        }
    }
    return true;
}

export default testWatchOrderBookForSymbols;

File: julia/Ccxt/test/pro/exchange/test.watchOrderBookForSymbols.jl
using Test
using Ccxt
function testWatchOrderBookForSymbols(exchange, skippedProperties, symbols)

    method = "watchOrderBookForSymbols";
    currentTime = milliseconds(exchange);
    deadline = currentTime + 15000;
    seenSymbols = [];
    while functions.ccxtruthy(@functions.ccxt_or(functions.ccxt_lt(currentTime, deadline), functions.ccxt_lt(length(seenSymbols), length(symbols))))
        response = nothing;
        succeeded = true;
        try
            response = Base.fetch(watchOrderBookForSymbols(exchange, symbols));
        catch e
            if functions.ccxtruthy(@functions.ccxt_and(!functions.ccxtruthy(isTemporaryFailure(testSharedMethods, e)), !functions.ccxtruthy((isa(e, InvalidNonce)))))
                throw(e);
            end
            currentTime = milliseconds(exchange);
            succeeded = false;

        end
        if functions.ccxtruthy(@functions.ccxt_and((succeeded), (response != nothing)))
            @test functions.ccxtruthy(isDictionary(exchange, response))
            currentTime = milliseconds(exchange);
            assertInArray(testSharedMethods, exchange, skippedProperties, method, response, "symbol", symbols);
            testOrderBook(exchange, skippedProperties, method, response, nothing);
            symbol = get(response, Symbol("symbol"), nothing);
            if functions.ccxtruthy(@functions.ccxt_and((symbol != nothing), !functions.ccxtruthy(inArray(exchange, symbol, seenSymbols))))
                                push!(seenSymbols, symbol);
            end
        end
    end
    return true
end
